src/main/java/org/osc/manager/ism/api/IsmAgentApi.java

```diff
@@ -16,79 +16,75 @@
+import org.osc.manager.ism.entities.DeviceMemberEntity;
+import org.osc.manager.ism.entities.DeviceMemberStatusEntity;
 import org.osc.sdk.manager.api.ManagerDeviceMemberApi;
 import org.osc.sdk.manager.element.ApplianceManagerConnectorElement;
 import org.osc.sdk.manager.element.DistributedApplianceInstanceElement;
 import org.osc.sdk.manager.element.ManagerDeviceMemberStatusElement;
 import org.osc.sdk.manager.element.VirtualSystemElement;
-/**
- * Only used for testing
- */
-public final class IsmAgentApi implements ManagerDeviceMemberApi {
+import org.osgi.service.transaction.control.TransactionControl;
 
-    private Logger log = Logger.getLogger(IsmAgentApi.class);
+@SuppressWarnings("unused")
+public final class IsmAgentApi implements ManagerDeviceMemberApi {
+    private static final Logger LOG = Logger.getLogger(IsmAgentApi.class);
+    private IsmDeviceApi api;
+    private final TransactionControl txControl;
+    private final EntityManager em;
+    private final VirtualSystemElement vs;
 
-    private IsmAgentApi(ApplianceManagerConnectorElement mc, VirtualSystemElement vs) throws Exception {
-        // Set mc as class variable as needed
-        this.log.info("Agent API created for Virtual System: " + vs.getName());
+    public IsmAgentApi(ApplianceManagerConnectorElement mc, VirtualSystemElement vs, TransactionControl txControl,
+            EntityManager em) throws Exception {
+        this.txControl = txControl;
+        this.em = em;
+        this.vs = vs;
+        this.api = new IsmDeviceApi(this.vs, this.txControl, this.em);
     }
 
-    public static IsmAgentApi create(ApplianceManagerConnectorElement mc, VirtualSystemElement vs) throws Exception {
-        return new IsmAgentApi(mc, vs);
+    public static IsmAgentApi create(ApplianceManagerConnectorElement mc, VirtualSystemElement vs,
+            TransactionControl txControl, EntityManager em) throws Exception {
+        return new IsmAgentApi(mc, vs, txControl, em);
     }
 
     @Override
     public List<ManagerDeviceMemberStatusElement> getFullStatus(List<DistributedApplianceInstanceElement> list) {
-        return getAgentFullStatus(list);
-    }
-
-    private List<ManagerDeviceMemberStatusElement> getAgentFullStatus(List<DistributedApplianceInstanceElement> list) {
-        // Get basic status
-        ManagerDeviceMemberStatusElement agentElem = null;
         List<ManagerDeviceMemberStatusElement> response = new ArrayList<>();
-        if (list != null){
-            for (DistributedApplianceInstanceElement dai: list){
-                agentElem = getAgentBasicStatus(dai);
-                response.add(agentElem);
+        if (list != null) {
+            for (DistributedApplianceInstanceElement dai : list) {
+                DeviceMemberEntity member = null;
+                try {
+                    member = (DeviceMemberEntity) this.api.findDeviceMemberByName(dai.getName());
+                } catch (Exception e) {
+                    e.printStackTrace();
```

Comment:
you should log error not print stack trace.

In this case, if the dai is not found we should throw the exception back to OSC.

Reply:
yes will take care.

Reply:
revert the `if` to avoid deep nesting.... if (list == null) { throw illegalargument...} the rest of the code comes after.